### ISSUE TITLE
Custom format of validation error response

### ISSUE THREAD

**Issue by JakubMatejka** [vandium-io/vandium-node#12]
Is it possible to change default format of validation error? 

E.g. I have this validation setup:

```
module.exports.login = vandium.createInstance({
  validation: {
    schema: {
      body: vandium.types.object().keys({
        email: vandium.types.email().required()
          .error(new UserError('Parameter email is required and should have ' +
          'format of email address')),
        password: vandium.types.string().required()
          .error(new UserError('Parameter password is required')),
      }),
    },
  },
})
```

And when I break it, API Gateway returns something like this

```
{
    "errorMessage": "validation error: Parameter email is required and should have format of email address",
    "errorType": "Error",
    "stackTrace": []
}
```

But I would like to change it, e.g. remove `stackTrace` completely.

In addition I am trying to adapt `lambda-proxy` integration using Serverless and API Gateway fails completely with internal server error and saying `Execution failed due to configuration error: Malformed Lambda proxy response` in the log.

**Comment by richardhyatt** [comment]
Few suggestions for API Gateway:

1. Use the Vandium configuration parameter `stringifyErrors: true`. This will force all exceptions to be converted into a string and stored in the "errorMessage" field. In addition, you can return more fields from the exceptions (and you'll have to set `stripErrors: false` to see them).

2. Use a response template to transform the message:

```
#set ($errorMessageObj = $util.parseJson($input.path('$.errorMessage')))
#set ($bodyObj = $util.parseJson($input.body))

{
#if( $errorMessageObj.errorMessage != "" )
    "type" : "$errorMessageObj.errorType",
    "message" : "$errorMessageObj.errorMessage",
#else
    "type": "error",
    "message": "$bodyObj.errorMessage"
#end
}
``` 

This mapping will decode the "stringified" exception from the errorMessage field OR work with the existing message if a timeout/process exit case.

**Comment by JakubMatejka** [comment]
Thanks for your suggestions but I'm not sure I can use them. We have the integration request of API Gateway set as lambda proxy with "Lambda Proxy integration". Which means that I cannot edit integration response and the result of lambda function must have this format: 
```
{
    "statusCode": httpStatusCode,
    "headers": { "headerName": "headerValue", ... },
    "body": "..."
}
```
(see http://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-set-up-simple-proxy.html#api-gateway-simple-proxy-for-lambda-output-format)

Unfortunately I can't change this APIG settings because it would break other things in our app. Do you support it or do you plan to support it anytime soon?

**Comment by richardhyatt** [comment]
We need to add this functionality - I will add this to the 3.3 branch (hope to release sometime next week)

**Comment by richardhyatt** [comment]
Would be interested in hearing how you think it should look from the user's viewpoint.

**Comment by JakubMatejka** [comment]
That's great, thank you. Well, there are two steps. First is simple and needs just one configuration parameter saying that Vandium should encapsulate current output with the required envelope (i.e. it would look like `{ "statusCode": x, "body": JSON.stringify({ "errorMessage": "validation error: ...", "errorType": "Error", "stackTrace": [] }), "headers": {} }`. Plus it would be nice to be able setup headers.

Second step is customization of the `body`. Our API formats errors like `{ "error": "message", "errorType": "validation", "requestId": "aws-request-id" }` and it is not consistent nor nice that Vandium formats them otherwise. But so far I'm not sure how the configuration should look like, I'm afraid that API Gateway won't help us like with response templates, right?

**Comment by JakubMatejka** [comment]
Well, maybe it could be simple too, what about something like:
```
{
  validation: {
    schema: {...},
    errorFormat: (msg) { 
      return JSON.stringify({ error: msg, errorType: "validation" });
    },
  }
}
```
Or the callback could return object which would be stringified by Vandium, I'm not sure what's better.

Probably I would not be able to put there aws request id anyway but it would be still much better then now. ;o)

**Comment by richardhyatt** [comment]
I've added support for the lambda proxy in the 3.x branch - several ways to use it (using your login example):

### simple proxy with no headers
```js
module.exports.login = vandium.createInstance({
  validation: {
    
     // as normal
  },

  lambdaProxy: true
});
```

### simple proxy with custom headers
```js
module.exports.login = vandium.createInstance({
  validation: {
    
     // as normal
  },

  lambdaProxy: {
    
    headers: {
      "x-my-header": "whatever"
    }
  }
});
```

### Custom implementation

```js
class MyLambdaProxy extends vandium.LambdaProxy {

  constructor() {
    
    super();
  }
 
  onError( event, context, error ) {

     let proxyObject = {};

     proxyObject.statusCode = error.status || error.statusCode || 500;
     proxyObject.headers = { };
 
     let body = {

       type: error.name,
       message: error.message,
       bad: true
    };

    proxyObject.body = JSON.stringify( body );
    return proxyObject;
  }
}

module.exports.login = vandium.createInstance({
  validation: {
    
     // as normal
  },

  lambdaProxy: new MyLambdaProxy()
});
```

Give it a try (`npm install vandium-io/vandium-node`) and let me know if it works for you.

**Comment by JakubMatejka** [comment]
It's looking great, thanks! I hope to get to it soon

**Comment by richardhyatt** [comment]
Will also be adding the ability to simplify validation when lambda proxy support is enabled.

Validation will look like:

```js
{
  validation: {
    schema: {
    
      GET: {
        // validation for GET requests
      },
      POST: {
        // validation for POST request
      }
    }
  },
  lambdaProxy: true
}
```

**Comment by richardhyatt** [comment]
Resolved in 3.3.0
